core/arch/arm/mm/pgt_cache.c

```diff
@@ -340,6 +341,128 @@ void pgt_flush_ctx(struct tee_ta_ctx *ctx)
 	mutex_unlock(&pgt_mu);
 }
 
+static void flush_pgt_entry(struct pgt *p)
+{
+	tee_pager_pgt_save_and_release_entries(p);
+	assert(!p->num_used_entries);
+	p->ctx = NULL;
+	p->vabase = 0;
+}
+
+static bool pgt_entry_matches(struct pgt *p, void *ctx, vaddr_t begin,
+			      vaddr_t last)
+{
+	if (!p)
+		return false;
+	if (p->ctx != ctx)
+		return false;
+	if (!core_is_buffer_inside(p->vabase, SMALL_PAGE_SIZE, begin,
+				   last - begin))
+		return false;
+
+	return true;
+}
+
+static void flush_ctx_range_from_list(struct pgt_cache *pgt_cache, void *ctx,
+				      vaddr_t begin, vaddr_t last)
+{
+	struct pgt *p;
+	struct pgt *next_p;
+
+	p = SLIST_FIRST(pgt_cache);
+	while (pgt_entry_matches(p, ctx, begin, last)) {
+		flush_pgt_entry(p);
+		SLIST_REMOVE_HEAD(pgt_cache, link);
+		push_to_free_list(p);
+		p = SLIST_FIRST(pgt_cache);
+	}
+
+	p = SLIST_FIRST(pgt_cache);
+	while (true) {
+		next_p = SLIST_NEXT(p, link);
+		if (!next_p)
+			break;
+		if (pgt_entry_matches(next_p, ctx, begin, last)) {
+			flush_pgt_entry(next_p);
+			SLIST_REMOVE_AFTER(p, link);
+			push_to_free_list(next_p);
+			continue;
+		}
+
+		p = SLIST_NEXT(p, link);
+	}
+}
+
+void pgt_flush_ctx_range(struct pgt_cache *pgt_cache, void *ctx,
+			 vaddr_t begin, vaddr_t last)
+{
+	mutex_lock(&pgt_mu);
+
+	flush_ctx_range_from_list(pgt_cache, ctx, begin, last);
+	flush_ctx_range_from_list(&pgt_cache_list, ctx, begin, last);
+
+	condvar_broadcast(&pgt_cv);
+	mutex_unlock(&pgt_mu);
+}
+
+void pgt_transfer(struct pgt_cache *pgt_cache, void *old_ctx, vaddr_t old_va,
+		  void *new_ctx, vaddr_t new_va, size_t size)
+{
+	const size_t pgtsize = CORE_MMU_PGDIR_SIZE;
+	const vaddr_t new_base = ROUNDDOWN(new_va, pgtsize);
+	const vaddr_t old_base = ROUNDDOWN(old_va, pgtsize);
+	const size_t num_new_pgt = (size - 1 + new_va - new_base) / pgtsize + 1;
```

Comment:
if `size == 0`,  then `num_new_pgt` will be some big value (check http://ideone.com/XID9h3)